Clean up and update to OptimizationBase@v4

SebastianM-C opened the pull request:
Since we always dispatch on the optimizer, it makes much more sense
to have this as the first argument (and first type parameter).

## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

Add any other context about the problem here.

ChrisRackauckas commented:
Make this a 4.0 and let's just do it. We are basically the only ones that depend on OptimizationBase directly so it's okay this didn't get lumped into the v3.

ChrisRackauckas commented:
We just need to merge, release OptimizationBase v4, then downstream test via sources, then merge and release stuff. But the fact that v1.12 came out and the autodiff packages red X our CI just absolutely murders our ability to check this stuff automatically, so 😅 it's a bit rough right now.

Let me make sure v2 backports are in a good spot first then we do this.

Also, maybe we might want an IIP dispatch on this sooner or later?

SebastianM-C commented:
so all the AD generated functions would have both variants and the loss is oop only?

SebastianM-C commented:
https://github.com/SciML/SciMLBase.jl/pull/1161

SebastianM-C commented:
##  Changes Summary

##  Core Interface Refactoring (2 commits)

  1. Reorder fields in OptimizationCache (f2282637)
    - Moved optimizer to first field/type parameter for cleaner dispatches
  2. Bump OptimizationBase compat (f39ccb2e)
    - Updated compatibility after cache field reordering

###  API Modernization (1 commit)

  3. Replace supports_opt_cache_interface with has_init (2786ea59)
    - Migrated all optimizers to use SciMLBase.has_init (requires SciMLBase 2.122+)

###  Test Infrastructure & Julia 1.11 Support (4 commits)

  4. Make OptimizationBase tests runnable on Julia v1.11 (b24295b7)
    - Updated test Project.toml and fixed compatibility issues
  5. Fix deprecations (7b7a143f)
  6. Make CI run for LTS (279ae963)
  7. Fix test env setup (d9c1eb80)

###  Package-Specific Test Fixes (9 commits)

  8. OptimizationPRIMA: Fix cache construction (e1645a8a)
  9. OptimizationSciPy: Fix tests (b732ced4)
  10. OptimizationPolyalgorithms: Fix test env (ed1db983)
  11. OptimizationNLPModels: Fix test env (7a485055)
  12. OptimizationMadNLP:
    - Use MadNLP@0.8.12 API (d33c6630)
    - Cleanup interface (9fd21df5)
    - Fix tests (70552a45)
  13. OptimizationManopt: Avoid errors and warnings in tests (9bbb4ec3)
  14. IncompatibleOptimizerError location fixes (2 commits: b2a7eb2b, becf556a)

Regarding the fixes for `OptimizationManopt` (13) we were getting errors in the tests because the implementation was trying to call functions that we not generated by AD, so I changed `SciMLBase.requireshessian` to `true`. I understood that this is not necessarily required from https://github.com/SciML/Optimization.jl/pull/1009, but if we don't change the implementation we to use the approximate formulations, we will get an error. Let me know if this needs to be reverted.

###  Documentation & Environment (2 commits)

  16. Fix docs env (7b8dfd06)
    - Updated OptimizationBase compat: "2, 3" → "4"
    - Updated SciMLBase compat: "2.30.0" → "2.122.1"
    - Added complete [sources] section with all 15+ lib subpackages
  17. Skip test_stale_deps on 1.10 (68396632)
    - Workaround for Julia 1.10 lacking [sources] support

###  Bug Fixes (1 commit)

  18. Fix Enzyme HVP returning zeros (b0860b8d)
    - Fixed Hessian-vector product returning [0, 0] instead of correct values
    - Root cause: Enzyme.make_zero(x) zeroing out tangent direction
    - Fixed both in-place and out-of-place implementations
